[aes] Add list of countermeasures in preparation for D2S sign-off

vogelpi opened the pull request:
This PR adds the list of countermeasures for AES in preparation for the D2S sign-off.

This requires #9816 to go in first.

vogelpi commented on hw/ip/aes/data/aes.hjson:
```diff
@@ -194,11 +209,78 @@
+    { name: "CONFIG.SPARSE",
+      desc: "Critical fields in main control register one-hot encoded."
+    }
+    { name: "DATA_REG.SW_UNREADABLE",
```
I wasn't sure whether to put `KEY` or `DATA_REG` here. `DATA_REG` stands for "non-software config data registers (e.g. keccak state)". Could also list the two separately. WDYT?

rswarbrick replied:
I guess maybe I'd split it, but don't have strong feelings.

vogelpi replied:
Sounds good, I've now split it up into to items for `KEY` and `DATA_REG`.

vogelpi commented on hw/ip/aes/doc/checklist.md:
```diff
@@ -75,8 +75,8 @@ Security      | [SEC_RND_CNST][]        | Done        |
 --------------|------------------------------|-------------|------------------
-Security      | [SEC_CM_IMPLEMENTED][]       | Not Started |
-Security      | [SEC_CM_ASSETS_LISTED][]     | Not Started |
+Security      | [SEC_CM_IMPLEMENTED][]       | Done        | Datapath FI hardening in progress (not strictly required)
```
There is still some review backlog for other AES changes. At the same, datapath FI hardening has never been a strict requirement for the first version. We've thus decided to take AES to D2S before datapath FI hardening is complete and continue the implementation in parallel.

rswarbrick commented on hw/ip/aes/data/aes.hjson:
```diff
@@ -194,11 +209,81 @@
+    { name: "KEY.MASKING",
+      desc: "1st-order domain-oriented masking of the cipher core."
+    }
+    //{ name: "DATA_REG.INTEGRITY",
+    //  desc: "Datapath FI hardening."
+    //}
```
Could you add a note here to explain why this is included but commented out? ("We're gonna do this, but haven't done it yet" or similar)

vogelpi replied:
Done.

rswarbrick commented on util/reggen/countermeasure.py:
```diff
@@ -31,6 +31,7 @@
     'DIFF',
     'REDUN',
     'SHADOW',
+    'MULTIRAIL'
```
Missing comma here!

rswarbrick replied:
Also, I think the only documentation for these names is in the (confidential) countermeasures sheet: is that right? If so, we should probably keep that in sync with this file.

vogelpi replied:
Thanks for spotting this. I've fixed that now.

rswarbrick replied:
About the documentation, I've also just opened https://github.com/lowRISC/opentitan/issues/9985, which tracks documenting these properly.

rswarbrick commented on util/reggen/countermeasure.py:
```diff
@@ -40,6 +41,7 @@
     'BKGN_CHK',
     'GLITCH_DETECT',
     'SW_UNREADABLE',
+    'SEC_WIPE'
```
Missing comma here!

rswarbrick reviewed (commented):
The failing CI is because of the missing commas in `countermeasure.py`

rswarbrick commented on hw/ip/aes/data/aes.hjson:
```diff
@@ -194,11 +209,81 @@
+      desc: "Key, IV and data input registers are write only for software."
+    }
+    { name: "DATA_REG.SEC_WIPE",
+      desc: "Key, IV, input/output data and internal state registers can be cleared with pseudo-random data."
```
Are there situations where they're cleared with something else? Otherwise, "are cleared" might be better?

rswarbrick commented on hw/ip/aes/data/aes.hjson:
```diff
@@ -194,11 +209,83 @@
+    { name: "DATA_REG.SEC_WIPE",
+      desc: "Key, IV, input/output data and internal state registers can be cleared with pseudo-random data."
+    }
+    { name: "MAIN.FSM.SPARSE",
```
I'm not sure that this is how the countermeasure instance name is supposed to be used. @msfschaffner: Is this supposed to be allowed? If not, I think there's a missing check in the Python :-)

That said, I think it's a good idea! If it's not currently how we intend things to work, maybe we should explicitly allow it? (In which case, I'd suggest allowing hierarchical instance names, so use `rsplit(..., 2)` in the Python and treat the stuff to the left as an opaque blob)

vogelpi replied:
`MAIN` is the optional instance name and `FSM` the asset. This is inline with what @msfschaffner proposed in #9816.

msfschaffner replied:
So the naming convention we outlined was to allow for an optional uniquifier in front of the canonical name, see also on the spreadsheet where the naming convention has initially been proposed.

I am not sure we should go full-on hierarchical with unbounded number uf hierarchy levels, though.
My concern is that this may just lead to more naming discussions, as it would allow for more naming combinations.

msfschaffner replied:
@tjaychen

tjaychen replied:
yeah the optional prefix was mostly a mechanism to allow the countermeasure writer a way to differentiate among instances of countermeasures.  There are not a lot of strict rules about this.  I would prefer not to go down the path of full hierarchy, because if we do, the question will become how we validate that.

rswarbrick commented on hw/ip/aes/data/aes.hjson:
```diff
@@ -194,11 +209,83 @@
+      desc: "The CTR mode FSM moves to a terminal error state upon local escalation."
+    }
+    { name: "CTR.REDUN",
+      desc: "The AES round counter is protected by a second counter counting in the opposite direction."
```
Nit: This would probably read slightly more easily with "... counter that counts in ... "

vogelpi replied:
Yep, I agree.

vogelpi commented:
Thanks @rswarbrick for your review.

Now CI is failing because I didn't rerun reggen. But this is only possible after merging #9816 .

msfschaffner commented on doc/rm/comportability_specification/index.md:
```diff
@@ -299,6 +299,8 @@ The following standardised countermeasures are defined:
 | GLITCH_DETECT  | The asset is protected by an analog glitch detector | CTRL, FSM, CLK, RST
 | SW_UNREADABLE  | The asset is not accessible by software |
+| SIDELOAD       | The asset is can be loaded without exposing it to software | KEY
+| SEC_WIPE       | The asset is initialized or cleared using pseudo-random data | KEY, DATA_REG, MEM
```
this one makes sense to me

msfschaffner commented on doc/rm/comportability_specification/index.md:
```diff
@@ -299,6 +299,8 @@ The following standardised countermeasures are defined:
 | BKGN_CHK       | The asset is protected with a continuous background health check |
 | GLITCH_DETECT  | The asset is protected by an analog glitch detector | CTRL, FSM, CLK, RST
 | SW_UNREADABLE  | The asset is not accessible by software |
+| SIDELOAD       | The asset is can be loaded without exposing it to software | KEY
```
isn't this similar to `SW_UNREADABLE`?

vogelpi commented:
Update: rebased on top of master and reworked the labels as discussed in the D2S review meeting. In particular:
- Label IV as seperate asset (IV.CONFIG.SW_UNREADABLE and IV.CONFIG.SEC_WIPE)
- Expand description of all FSM.LOCAL_ESC to indicate what can trigger the escalation. Related to that, remove non-FSM-specific LOCAL_ESC labels. One exemption is DATA_REG.LOCAL_ESC: I think it makes sense to indicate that we don't release intermediate data upon escalation.
- Make KEY.SCA more specific (DATA_REG.KEY.SCA).
- Add LC_ESCALATE_EN.INTERSIG.MUBI label.

msfschaffner reviewed (approved):
LGTM, thanks for updating this.

cdgori commented:
When this lands we (I) have to update the spreadsheet, I think?

vogelpi commented:
> When this lands we (I) have to update the spreadsheet, I think?

Yes, this is my understanding, too. I don't have write access to the spreadsheet.

cdgori commented:
> > When this lands we (I) have to update the spreadsheet, I think?
> 
> Yes, this is my understanding, too. I don't have write access to the spreadsheet.

OK, I have at least gotten everything from the HJSON into the spreadsheet and collapsed/expanded entries where I could.  I need some help (from you, I think?) to resolve the ones at the bottom of the AES section that link to #1005 (don't worry about the REJECTED ones, they will carry over fine).  I'm not sure if: they are still to-be-done (or consciously not-to-be-done and thus converted to REJECTED), could be merged into other entries, or require another countermeasure entry in the HJSON.  If you have time after tomorrow's Si meeting we might be able to finish it off then?
